src/liger_kernel/ops/backends/_ascend/ops/rope.py

```diff
@@ -0,0 +1,229 @@
+        cos_bs == 1,
+        cos_row_idx * sin_row_stride,
+        batch_idx * (sl * sin_row_stride) + cos_row_idx * sin_row_stride,
+    )
+    d_idx = tl.arange(0, hd // 2)
+    cos_vals = tl.load(cos + d_idx)
+    sin_vals = tl.load(sin + d_idx)
+
+    q_base = q_ptr + pid * q_row_stride
+    k_base = k_ptr + pid * k_row_stride
+
+    # Process in chunks to prevent UB overflow
+    for qh_block in range(0, n_qh, BLOCK_Q):
+        qh_idx = tl.arange(0, BLOCK_Q) + qh_block
+
+        qh_mask = qh_idx < n_qh
+        block_mask = qh_mask[:, None]
+
+        offsets = qh_idx[:, None] * hd + d_idx[None, :]
+
+        q_left = tl.load(q_base + offsets, mask=block_mask, other=0)
+        q_right = tl.load(q_base + offsets + (hd // 2), mask=block_mask, other=0)
+
+        if not BACKWARD_PASS:
+            new_left = q_left * cos_vals - q_right * sin_vals
+            new_right = q_right * cos_vals + q_left * sin_vals
+        else:
+            new_left = q_left * cos_vals + q_right * sin_vals
+            new_right = q_right * cos_vals - q_left * sin_vals
+
+        tl.store(q_base + offsets, new_left, mask=block_mask)
+        tl.store(q_base + offsets + (hd // 2), new_right, mask=block_mask)
+
+    for kh_block in range(0, n_kh, BLOCK_K):
+        kh_idx = tl.arange(0, BLOCK_K) + kh_block
+
+        kh_mask = kh_idx < n_kh
+        block_mask = kh_mask[:, None]
+
+        offsets = kh_idx[:, None] * hd + d_idx[None, :]
+
+        k_left = tl.load(k_base + offsets, mask=block_mask, other=0)
+        k_right = tl.load(k_base + offsets + (hd // 2), mask=block_mask, other=0)
+
+        if not BACKWARD_PASS:
+            new_left = k_left * cos_vals - k_right * sin_vals
+            new_right = k_right * cos_vals + k_left * sin_vals
+        else:
+            new_left = k_left * cos_vals + k_right * sin_vals
+            new_right = k_right * cos_vals - k_left * sin_vals
+
+        tl.store(k_base + offsets, new_left, mask=block_mask)
+        tl.store(k_base + offsets + (hd // 2), new_right, mask=block_mask)
+
+def calculate_optimal_block_size(
+    dtype_size_bytes: int,
+    hd: int,
+    n_heads: int,
+    safety_factor: float = 0.9 # Reserve 10% space
+) -> int:
+    dev_props = torch.npu.get_device_properties(0).name
+    tbe.common.platform.set_current_compile_soc_info(dev_props)
+    ub_size_bytes=tbe.common.platform.get_soc_spec("UB_SIZE")
+
+    max_elements = int(ub_size_bytes / dtype_size_bytes * safety_factor)
+
+    # total ub cost:
+    # cos_vals, sin_vals, d_idx: hd//2
+    # qh_idx, qh_mask: BLOCK_Q
+    # block_mask, offsets, q_left, q_right, new_left, new_right: BLOCK_Q × (hd//2)
+
+    # 3 * hd // 2 + (3 * hd + 2) * max_block <= max_elements
+    max_block = min(
+        n_heads,
+        int((max_elements - 3 * hd // 2) // (3 * hd + 2))
+    )
+
+    if max_block != triton.next_power_of_2(max_block):
+        return triton.next_power_of_2(max_block) // 2
+    else:
+        return triton.next_power_of_2(max_block)
```

Comment:
This part is a little bit confusing. 

Is total ub cost an assumption? or real data collected from profiling system? I'm not familiar with npu so I might be wrong, but in cuda I believe only q and k blocks are loaded into shared memory while others are just register values.

Reply:
Assuming the cost is correct, I suggest rewriting conditions for readibility.
```suggestion
    # 3 * hd // 2 + (3 * hd + 2) * max_block <= max_elements
    max_block = int((max_elements - 3 * hd // 2) // (3 * hd + 2))
    if max_block < n_heads:
        return triton.next_power_of_2(max_block) // 2
    else:
        return triton.next_power_of_2(n_heads) 
```